src/optimizers/anyprecision_optimizer.py

```diff
@@ -0,0 +1,182 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+# AnyPrecisionAdamW: a flexible precision AdamW optimizer
+# with optional Kahan summation for high precision weight updates.
+# Allows direct control over momentum, variance and auxiliary compensation
+# buffer dtypes.
+# Optional Kahan summation is used to offset precision reduction for
+# the weight updates. This allows full training in BFloat16 (equal or
+# better than FP32 results in many cases) due to high precision weight upates.
+
+import torch
+from torch.optim.optimizer import Optimizer
+
+
+class AnyPrecisionAdamW(Optimizer):
+    def __init__(
+        self,
+        params,
+        lr=1e-3,
+        betas=(0.9, 0.999),
+        eps=1e-8,
+        weight_decay=0.0,
+        use_kahan_summation=False,
+        momentum_dtype=torch.float32,
+        variance_dtype=torch.bfloat16,
+        compensation_buffer_dtype=torch.bfloat16,
```

Comment:
this can be done in a follow up PR, but let's maybe consider not defaulting things to torch.bfloat16 eventually. this is because it might be good to make this optimizer usable out of the box with the defaults on all HW architectures, but only A100 supports bfloat16 well at the moment. 

But the downside here would be that the default optimizer won't be too interesting, it'd just be AdamW

Reply:
as an alternative - we could add in a simple bfloat16 support check, and then update the default for variance from BF16 to FP32 automatically (and optionally display a simple note to the user that bf16 native support not detected so reverting to FP32?)
This would allow it to work out of the box on any system while taking advantage of bfloat automatically when/where it can. 
Drawback is even a single line warning * X number of GPU's would not be great.

Reply:
added this enhancement issue for tracking/discussion:
https://github.com/pytorch/torchdistx/issues/59

Reply:
actually I think having this concept of 'smart defaults' where it attempts appropriate BF16 but rolls back to FP32 when not supported is a nice user experience. Could also apply for kahan summation being turned on and we revert it back to off, if BF1t6 is not supported and enable both momentum and variance to FP32 automatically (since Kahan adds no value for FP32).  This would be nice as they would inherently get the optimal setup on various hardware. 
That also gives it a bit more value even in the scenario of it becomes AdamW b/c it also reverts automatically on future hardward where BF16 is supported.